subxt/examples/reconnecting_rpc_client.rs

```diff
@@ -0,0 +1,103 @@
+        // then disconnect.
+        //
+        // This is just a way to ensure that the connection isn't idle if no message is sent that often
+        .enable_ws_ping(
+            PingConfig::new()
+                .ping_interval(Duration::from_secs(6))
+                .inactive_limit(Duration::from_secs(30)),
+        )
+        // There are other configurations as well that can be found here:
+        // <https://docs.rs/reconnecting-jsonrpsee-ws-client/latest/reconnecting_jsonrpsee_ws_client/struct.ClientBuilder.html>
+        .build("ws://localhost:9944".to_string())
+        .await?;
+
+    let api: OnlineClient<PolkadotConfig> =
+        OnlineClient::from_rpc_client(RpcClient::new(rpc.clone())).await?;
+
+    // Build a balance transfer extrinsic.
+    let dest = dev::bob().public_key().into();
+    let balance_transfer_tx = polkadot::tx().balances().transfer_allow_death(dest, 10_000);
+
+    // Submit the balance transfer extrinsic from Alice, and wait for it to be successful
+    // and in a finalized block. We get back the extrinsic events if all is well.
+    let from = dev::alice();
+
+    let mut balance_transfer_progress = api
+        .tx()
+        .sign_and_submit_then_watch_default(&balance_transfer_tx, &from)
+        .await?;
+
+    // When the connection is lost, a error RpcError::DisconnectWillReconnect is emitted on the stream.
+    // in such scenarios the error will be seen here.
+    //
+    // In such scenario it's possible that messages are lost when reconnecting
+    // and if that's acceptable you may just ignore that error message.
+    while let Some(status) = balance_transfer_progress.next().await {
+        match status {
+            // It's finalized in a block!
+            Ok(TxStatus::InFinalizedBlock(in_block)) => {
+                // grab the events and fail if no ExtrinsicSuccess event seen:
+                let events = in_block.wait_for_success().await?;
+                // We can look for events (this uses the static interface; we can also iterate
+                // over them and dynamically decode them):
+                let transfer_event = events.find_first::<polkadot::balances::events::Transfer>()?;
+
+                if transfer_event.is_some() {
+                    println!("Balance transfer success");
+                } else {
+                    println!("Failed to find Balances::Transfer Event");
+                }
+            }
+            // Just log any other status we encounter:
+            //
+            // In this example we emit some important status handling for
+            // here such as Dropped, Invalid etc....
+            Ok(_) => {
+                println!("New status");
+            }
+            // In this example we just ignore when reconnections occurs
+            // but it's technically possible that we can lose
+            // messages on the subscription such as `InFinalizedBlock`
```

Comment:
I like this example overall, but when submitting a tx it's particularly tricky to handle as you will lose the subscription anyway won't you and won't get any new messages after reconnecting?

So perhaps a better (and simpler, anyway) example is just subscribing to finalized blocks?

I can't remember how your reconnecting client handles subscriptions etc; does it automatically re-subscribe behind the scenes? That area is an interesting one because on one hand, it might be better if the reconnecting client just reconnects but doesn't try to re-establish anything (so that the Subxt backend's can choose how best to restart stuff), but on the other hand having the client do it is simpler at least (but I fear may not always do what you want). Ultimately we'd need to agree on a "standard" for what is expected from a reconnecting client when it's used in Subxt

Reply:
> I can't remember how your reconnecting client handles subscriptions etc; does it automatically re-subscribe behind the scenes?

Yes, it re-subscribes to subscriptions and re-transmits pending method calls. 

It's just a PITA if one have plenty of storage subscriptions opened and manually have to re-subscribe to them.
For rpc-v2 then it's a different story but not really thought of in this use-case.

Reply:
Let's just make it optional and folks can try it out and complain what's bad :)

Reply:
Yeahh for now it's behind a feature flag so we can experiment!

My gut feeling is that before it becomes "stable" in Subxt, we'll prob end up wanting the Backend impls to decide when/how to resubscribe etc themselves (eg maybe for TXs we don't try re-sending but for following new/final blocks we just restart it or whatever

Reply:
Alex had a few cool ideas about distinguish between "stateful (submit_and_watch)" and "stateless (subscribe_storage)" subscriptions but still I think this reconnecting rpc client won't really work that well with rpc-v2 spec.

One would need to call a few methods for the `chainHead_unstable_follow` to make emit interesting stuff..

Reply:
> So perhaps a better (and simpler, anyway) example is just subscribing to finalized blocks?

I want to show how to ignore "DisconnectWillReconnect` and continue if a reconnection occurs. If you just call `wait_for_finalized_success` will quit with error but we could `wait_for_finalized_success_and_reconnect`... I guess.

Reply:
I think this conversation is good start to open another issue how to deal with it reconnections in subxt properly especially with the rpc v2 spec in mind which is particular tricky.

In general, I think just reconnect is probably the way to go and not re-subscribe and so on but we let's try my crate first and collect some feedback what's the best way forward :)

Reply:
> I want to show how to ignore "DisconnectWillReconnect" and continue if a reconnection occurs. If you just call wait_for_finalized_success it will quit with that error but we could wait_for_finalized_success_and_reconnect... I guess.

What I mean is just that it's less code to subscribe to finalied blocks and then show how you can ignore the DisconnectedWillReconnect error. Probably closer to being a valid thing that we can handle nicely too (we can just start the subscription up again as your client does atm anyway)

With transactions we probably(?) don't want to re-submit automatically so showing it as an example sortof hints at a behaviour that I don't think we'll keep, but anyway it's also just more effort building the tx etc when the point of the example is just handling the DisconnectWillReconnect error :)

> In general, I think just reconnect is probably the way to go and not re-subscribe and so on but we let's try my crate first and collect some feedback what's the best way forward :)

Yup I agree; I'm happy with this for now since it's behind a feature flag, but before we stabilise it I thnk we should make the RPC client not try to do anything clever and _only_ reconnect. Instead we can make the backend impls decide how to manage disconnects in followup PRs

Reply:
> What I mean is just that it's less code to subscribe to finalied blocks and then show how you can ignore the DisconnectedWillReconnect error. Probably closer to being a valid thing that we can handle nicely too (we can just start the subscription up again as your client does atm anyway)

> With transactions we probably(?) don't want to re-submit automatically so showing it as an example sortof hints at a behaviour that I don't think we'll keep, but anyway it's also just more effort building the tx etc when the point of the example is just handling the DisconnectWillReconnect error :)

Got it, makes sense